src/core/net/ip6_mpl.cpp

```diff
@@ -105,38 +102,265 @@ ThreadError Mpl::ProcessOption(const Message &aMessage, const Address &aAddress,
 
     VerifyOrExit(entry != NULL, error = kThreadError_Drop);
 
-    aForward = true;
+    entry->SetSeedId(aSeedId);
+    entry->SetSequence(aSequence);
+    entry->SetLifetime(kSeedEntryLifetime);
+    mSeedSetTimer.Start(kSeedEntryLifetimeDt);
+
+exit:
+    return error;
+}
+
+void Mpl::UpdateBufferedSet(uint16_t aSeedId, uint8_t aSequence)
+{
+    int8_t diff;
+    MplBufferedMessageMetadata messageMetadata;
+
+    Message *message = mBufferedMessageSet.GetHead();
+    Message *nextMessage = NULL;
+
+    // Check if multicast forwarding is enabled.
+    VerifyOrExit(GetTimerExpirations() > 0,;);
+
+    while (message != NULL)
+    {
+        nextMessage = message->GetNext();
+        messageMetadata.ReadFrom(*message);
+
+        if (messageMetadata.GetSeedId() == aSeedId)
+        {
+            diff = static_cast<int8_t>(aSequence - messageMetadata.GetSequence());
+
+            if (diff > 0)
+            {
+                // Stop retransmitting MPL Data Message that is consider to be old. The new
+                // MPL Data Message will be added after function exits.
+                mBufferedMessageSet.Dequeue(*message);
+                message->Free();
+            }
+
+            break;
+        }
+
+        message = nextMessage;
+    }
+
+exit:
+    return;
+}
+
+void Mpl::AddBufferedMessage(Message &aMessage, uint16_t aSeedId, uint8_t aSequence, bool forward)
+{
+    uint32_t now = Timer::GetNow();
+    ThreadError error = kThreadError_None;
+    MplBufferedMessageMetadata messageMetadata;
+    Message *messageCopy = NULL;
+    uint8_t hopLimit;
+    uint32_t nextTransmissionTime;
+
+    VerifyOrExit((messageCopy = aMessage.Clone()) != NULL, error = kThreadError_NoBufs);
+
+    if (forward)
+    {
+        aMessage.Read(Header::GetHopLimitOffset(), Header::GetHopLimitSize(), &hopLimit);
+        VerifyOrExit(--hopLimit > 0, error = kThreadError_Drop);
+        messageCopy->Write(Header::GetHopLimitOffset(), Header::GetHopLimitSize(), &hopLimit);
+    }
+
+    messageMetadata.SetSeedId(aSeedId);
+    messageMetadata.SetSequence(aSequence);
+    messageMetadata.SetTransmissionCount(forward ? 0 : 1);
+
+    // Emulate Trickle timer behavior and set up the next retransmission within [0,I] range.
+    messageMetadata.SetTransmissionTime(now + (otPlatRandomGet() % (kDataMessageInterval + 1)));
```

Comment:
nit: I don't think the `+ 1` is necessary.  Thread defines the interval as `[0, I)`.